docs/enhancements/20190724-prometheus-refactoring.md

```diff
@@ -0,0 +1,370 @@
+  - https://github.com/kubernetes/ingress-nginx/issues/3713
+  - https://github.com/kubernetes/ingress-nginx/pull/4139
+  - https://github.com/kubernetes/ingress-nginx/issues/2924
+  - https://github.com/kubernetes/ingress-nginx/issues/3898
+  - https://github.com/kubernetes/ingress-nginx/pull/4139
+
+### Non-Goals
+
+Update of the Grafana dashboard [provided by the project](https://github.com/kubernetes/ingress-nginx/tree/master/deploy/grafana/dashboards). This should be done after the implementation of this KEP.
+
+## Proposal
+
+Test if using message pack we can improve the performance of the Lua communication with the ingress controller ([lua library](https://luarocks.org/modules/antirez/lua-cmsgpack))
+
+One of the issues we have is the range of options and configuration users expected from the exposed metrics. This creates a maintenance problem due to the fact the metrics are hardcoded in the source code.
+For this reason, one of the goals in this KEP is to introduce a CRD for metrics, allowing users to define which rules they want to use and which labels should be used.
+As an example:
+
+```yaml
+kind: metric
+metadata:
+  name: request_duration_seconds
+  namespace: default
+spec:
+  help: The request processing time in milliseconds
+  constLabels:
+    - controller_pod
+  labels:
+    - ingress_uid # label to join with the ingress info metric using group_left
+    - path        # same ^^
+    - method
+    - status
+  type: counter|histogram|summary|gauge
+  buckets:
+    type: linear|exponential
+    values:
+      - 10
+      - 10
+      - 10
+```
+
+The default deployment should provide the same metrics we have now using this new `metric` type, using the configuration configmap to set the list of metrics to use, allowing users to define new types
+This approach allows us to also provide other types defined in other projects like VTS module or prometheus collector from nginx inc https://github.com/nginxinc/nginx-prometheus-exporter
+
+Use info metrics to reduce the number of labels (even constants) in the metrics. This will decrement the size of the metrics and just doing one join we can get the same cardinality.
+
+### Implementation Details/Notes/Constraints
+
+### Info metrics
+
+#### *Controller*
+
+- controller_class
+- controller_namespace
+- controller_pod
+
+#### *Ingresses* (only the leader handle this metrics)
+
+|Field name|Description|
+| --- | --- |
+|uid| Unique identifier of the ingress. This field is the key for the joins|
+|name| Name of the ingress|
+|namespace| Namespace where the ingress and the service are located|
+|class| Value of the annoation kubernetes.io/ingress.class|
+|host| Hostname of the ingress|
+|path| Path of the `http` backend in the ingress|
+|service| Name of the service being exposed|
+
+### Metrics
+
+**Name:** nginx_ingress_controller_bytes_sent
+
+**Description:** The the number of bytes sent to a client
+
+**Type:** histogram
+
+**Labels:**
+
+- ingress_uid
+- method
+- status
```

Comment:
An additional issue #3818 can be solved by incorporating a status_class:5xx in the metric implemented as part of this proposal